File: packages/goodomics/src/goodomics/schemas/field_references.py
"""Canonical references for analytical and project-metadata fields."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Annotated, Literal, cast

from pydantic import AfterValidator

MetadataEntity = Literal["subject", "sample", "run", "file"]
METADATA_ENTITIES: frozenset[str] = frozenset({"subject", "sample", "run", "file"})


def validate_field_reference_segment(value: str) -> str:
    """Require one non-empty, unescaped path segment for a stable field ID."""

    if not value or value != value.strip():
        raise ValueError(
            "Field-reference IDs must be non-empty and have no surrounding whitespace."
        )
    if "/" in value:
        raise ValueError("Field-reference IDs cannot contain '/'.")
    return value


FieldReferenceSegment = Annotated[str, AfterValidator(validate_field_reference_segment)]


@dataclass(frozen=True)
class ParsedFieldReference:
    """Resolved source identity and source-local field ID."""

    kind: Literal["contract", "metadata"]
    field_id: str
    contract_id: str | None = None
    entity: MetadataEntity | None = None


def contract_field_reference(contract_id: str, field_id: str) -> str:
    """Build the canonical unprefixed reference for one analytical field."""

    return (
        f"{validate_field_reference_segment(contract_id)}/"
        f"{validate_field_reference_segment(field_id)}"
    )


def metadata_field_reference(entity: str, field_id: str) -> str:
    """Build the canonical namespaced reference for one metadata field."""

    if entity not in METADATA_ENTITIES:
        raise ValueError(f"Unknown metadata entity: {entity!r}.")
    return f"metadata/{entity}/{validate_field_reference_segment(field_id)}"


def parse_field_reference(value: str) -> ParsedFieldReference:
    """Parse the strict canonical field-reference grammar."""

    if not value or value != value.strip():
        raise ValueError(
            "Field reference must be non-empty and have no surrounding whitespace."
        )
    parts = value.split("/")
    if len(parts) == 2:
        contract_id, field_id = parts
        return ParsedFieldReference(
            kind="contract",
            contract_id=validate_field_reference_segment(contract_id),
            field_id=validate_field_reference_segment(field_id),
        )
    if len(parts) == 3 and parts[0] == "metadata":
        _, entity, field_id = parts
        if entity not in METADATA_ENTITIES:
            raise ValueError(f"Unknown metadata entity in field reference: {entity!r}.")
        return ParsedFieldReference(
            kind="metadata",
            entity=cast(MetadataEntity, entity),
            field_id=validate_field_reference_segment(field_id),
        )
    raise ValueError(
        "Field reference must be '<contract>/<field>' or 'metadata/<entity>/<field>'."
    )

File: packages/goodomics/dashboard/src/lib/fieldReferences.ts
/** the purpose of this module is to provide utilities for creating and parsing field references. Some examples:
 `contractFieldReference("contract-123", "field-456")` returns `"contract-123/field-456"`
 `metadataFieldReference("subject", "age")` returns `"metadata/subject/age"`
*/

const METADATA_ENTITIES = new Set(["subject", "sample", "run", "file"]);

export type ParsedFieldReference =
  | { kind: "contract"; contractId: string; fieldId: string }
  | { kind: "metadata"; entity: string; fieldId: string };

function validateSegment(value: string) {
  if (!value || value !== value.trim() || value.includes("/")) {
    throw new Error(
      "Field-reference IDs must be non-empty and cannot contain '/'.",
    );
  }
  return value;
}

export function contractFieldReference(contractId: string, fieldId: string) {
  return `${validateSegment(contractId)}/${validateSegment(fieldId)}`;
}

export function metadataFieldReference(entity: string, fieldId: string) {
  if (!METADATA_ENTITIES.has(entity))
    throw new Error(`Unknown metadata entity: ${entity}.`);
  return `metadata/${entity}/${validateSegment(fieldId)}`;
}

export function parseFieldReference(value: string): ParsedFieldReference {
  if (!value || value !== value.trim())
    throw new Error("Invalid field reference.");
  const parts = value.split("/");
  if (parts.length === 2) {
    return {
      kind: "contract",
      contractId: validateSegment(parts[0]),
      fieldId: validateSegment(parts[1]),
    };
  }
  if (
    parts.length === 3 &&
    parts[0] === "metadata" &&
    METADATA_ENTITIES.has(parts[1])
  ) {
    return {
      kind: "metadata",
      entity: parts[1],
      fieldId: validateSegment(parts[2]),
    };
  }
  throw new Error(
    "Field reference must be '<contract>/<field>' or 'metadata/<entity>/<field>'.",
  );
}

export function isMetadataFieldReference(value: string) {
  try {
    return parseFieldReference(value).kind === "metadata";
  } catch {
    return false;
  }
}

export function valueReference(value: { field: string; as?: string }) {
  return value.as || value.field;
}
